### PR TITLE
Parse UPDATE/INSERT within WITH clause

### LINKED ISSUE #1888: CTEs before INSERT or UPDATE throws parsing errors
Getting parsing errors when defining CTEs before INSERT or UPDATE

## Expected Behaviour
No errors parsing something like
```
-- WITH before UPDATE
WITH mycte AS (SELECT * FROM mytable1)

UPDATE
    mytable2 m
SET
    column2 = column1
FROM
    mycte;

-- WITH before INSERT
WITH mycte AS (SELECT * FROM mytable1)

INSERT INTO table2 (column1, column2, column3)
VALUES ('value1', 'value2', 'value3');
```

## Observed Behaviour

```
sqlfluff lint with-sample.sql --ignore linting --dialect ansi
== [with-sample.sql] FAIL
L:   2 | P:   1 |  PRS | Line 2, Position 1: Found unparsable section: 'WITH
                       | mycte AS (SELECT * FROM mytable1)\n\n...'
L:  12 | P:   1 |  PRS | Line 12, Position 1: Found unparsable section: 'WITH
                       | mycte AS (SELECT * FROM mytable1)\n\n...'
```

## Steps to Reproduce
`sqlfluff lint with-sample.sql --ignore linting --dialect ansi`
## Dialect
ansi

## Version
sqlfluff, version 0.8.1
Python 3.9.2

## Configuration
default

### PR DESCRIPTION
**Pull request by jpy-git**
### Brief summary of the change made

PR fixes #1888. This PR adds the ability to parse INSERT/UPDATE within a WITH clause as described in the linked issue.

I have this working for the cases described in the linked issue, however, I am struggling to get the following valid query to parse:
```SQL
WITH mycte AS (
    SELECT
        foo,
        bar,
        baz
    FROM mytable1
)

INSERT INTO table2 (column1, column2, column3)
SELECT foo, bar, baz
FROM mycte;
```

I believe the dialect changes I made were sensible and should cover the above case, but still the parsing error, so any help/advice would be greatly appreciated! Thanks 😄 

### Are there any other side effects of this change that we should be aware of?
No, but need help resolving one case of not parsing before we can merge.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/dialects/ansi/insert_with_statement.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+INSERT INTO table2 (column1, column2, column3)
```
Comment (WittierDinosaur) [inline test/fixtures/dialects/ansi/insert_with_statement.sql:1]: You are missing a ; here - which is mangling this file

Comment (WittierDinosaur) [inline test/fixtures/dialects/ansi/insert_with_statement.sql:1]: Shouldn't the INSERT be after the CTE?

File: test/fixtures/dialects/ansi/with_insert_statement_b.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# computed by SQLFluff when running the tests. Please run
+# `python test/generate_parse_fixture_yml.py`  to generate them after adding or
+# altering SQL files.
+_hash: 4e55be927c4b974020c6f1ad74a4788a3944c840e8ac30e8d9a64c342d9c2e5d
+file:
+  statement:
+    with_compound_statement:
+      unparsable:
+      - keyword: WITH
+      - raw: mycte
```
Comment (WittierDinosaur) [inline test/fixtures/dialects/ansi/with_insert_statement_b.yml:12]: I think the behaviour you're seeing here will be resolved by the merge of #1894

Comment (WittierDinosaur) [reply test/fixtures/dialects/ansi/with_insert_statement_b.yml:12]: @jpy-git the fix has been merged now - so pulling in main may make this problem go away

File: test/fixtures/dialects/ansi/insert_with_statement.sql (hunk 1 above)
Comment (jpy-git) [reply test/fixtures/dialects/ansi/insert_with_statement.sql:1]: So this was already in the logic prior to this change, Bbut interestingly this is correct behaviour. Was new to me too 😄 

Postgres (from https://www.postgresql.org/docs/9.5/sql-insert.html):
image

Similar for Snowflake (https://community.snowflake.com/s/question/0D50Z00008EM1Qu/does-snowflake-support-ctewith-clause-on-an-insert-statement):
image

MySQL too! (https://stackoverflow.com/questions/24008316/insert-into-from-cte):
image

So I assume the decision was made in the past to include with the ANSI dialect 👍

File: test/fixtures/dialects/ansi/with_insert_statement_b.yml (hunk 2 above)
Comment (jpy-git) [reply test/fixtures/dialects/ansi/with_insert_statement_b.yml:12]: @WittierDinosaur amazing thank you, I'll take a look now!

File: test/fixtures/dialects/ansi/insert_with_statement.sql (hunk 1 above)
Comment (WittierDinosaur) [reply test/fixtures/dialects/ansi/insert_with_statement.sql:1]: Fair enough - makes sense!

**Review by WittierDinosaur — APPROVED** [review]

**Pull request merged**